WordPress/Classes/Services/PostCoordinator.swift

```diff
@@ -316,9 +315,11 @@ class PostCoordinator: NSObject {
         guard let context = post.managedObjectContext else {
             return
         }
+
         context.perform {
             if status == .failed {
                 self.mainService.markAsFailedAndDraftIfNeeded(post: post)
+                self.dispatchNotice(for: post)
```

Comment:
Not sure if I agree with this. The `change()` method originally was supposed to manage transitions between statuses. I wasn't even sure about `self.mainService.markAsFailedAndDraftIfNeeded()` in here. 

Adding a UI call is a different story though. 😐 This method can potentially be used anywhere we need a `status` transition and it can be misleading and surprising that it also shows a notice. 

Would you mind explaining more why we have to do it this way? Is there another way?

Reply:
You're right.

Here's why: when you have a failed image in a post the `upload` method will never be called in `save`. Thus, `upload`'s failure block will also never be reached. This means that there's no snackbar message in this case.

I've added it to `change()` because of that, to fill this gap. I can add this call in all the places the post is changed to `failed` in `prepareToSave` but then we're back to the original problem.

Thoughts?

Reply:
If we only need to handle `PostCoordinator.save()`, I think we can get away with adding the dispatch call there:

```diff
diff --git a/WordPress/Classes/Services/PostCoordinator.swift b/WordPress/Classes/Services/PostCoordinator.swift
index 9f90895da1..21122d2d5e 100644
--- a/WordPress/Classes/Services/PostCoordinator.swift
+++ b/WordPress/Classes/Services/PostCoordinator.swift
@@ -56,9 +56,16 @@ class PostCoordinator: NSObject {
             case .success(let post):
                 self.upload(post: post, completion: completion)
             case .error(let error):
-                if let notice = defaultFailureNotice {
-                    ActionDispatcher.dispatch(NoticeAction.post(notice))
+                switch error {
+                case SavingError.mediaFailure:
+                    let model = PostNoticeViewModel(post: postToSave)
+                    ActionDispatcher.dispatch(NoticeAction.post(model.notice))
+                default:
+                    if let notice = defaultFailureNotice {
+                        ActionDispatcher.dispatch(NoticeAction.post(notice))
+                    }
                 }
+
                 completion?(.error(error))
             }
         }
```

☝️ This code only _kinda_ works. The message is not correct but is probably worth it to look deeper into?

Reply:
I've tried that. Here and in `PostEditor+Publish` — which I think makes even more sense, since will get rid of UI stuff in `PostCoordinator`.

However, this doesn't work because of... synchronism (hello darkness). By the time this is called, `post.status` is still `pushing` and it's not yet marked to be auto uploaded. Thus, we always end up with "Post failed to upload" message. :( 

I can look deeper into it but this will probably lead us back to the sync/async stuff. I'm not sure if we wanna do that. 😅

Reply:
Hmm. 🤔 I wonder if we can add a completion block to `change()`. It's a bit ugly but maybe it'll work? 

```swift
                    self.change(post: post, status: .failed) {
                        completion(.error(SavingError.mediaFailure))
                    }
```

It seems low risk (famous last words).

Reply:
@shiki I refactored to allow the `change()` method to execute a completion block.

It's still not something that I'm proud of but this prevents unwanted behaviors in `change()`. Let me know what you think.